### PR TITLE
Progressive_view, batch_sizes can also change with resolution_milestones

### PR DESCRIPTION
**Pull request by voletiv**
1. Makes progressive increase of elevation, azimuth

Just set `progressive_until` steps in the config file. `elevation_range`, `azimuth_range` are progressively (linearly) increased from their default eval values to the full range values over `progressive_until` steps. If `progressive_until = 0`, no progression is made.

2. Can set `batch_size` as a list for different resolutions. For e.g.,
```
height: [64, 512]
width: [64, 512]
batch_size: [12, 3]
resolution_milestones: [1000]
```
In this case,
before iter1000, training is at 64x64 with batch_size 12,
after iter1000, training is at 512x512 with batch_size 3.

### CODE REVIEW

**Comment by claforte** [comment]
Thanks Vikram! 

This introduces a new convention and doesn't provide much control over the rate of expansion. Can't we instead reuse the existing convention, e.g. `min_step_percent: [0, 0.4, 0.02, 2000]  # (start_iter, start_val, end_val, end_iter)`?

**Comment by claforte** [comment]
BTW I'm not sure when radius or fov expansion would be desirable, and if so, whether we'd start from the larger value, or the smaller one. Either way, if we reuse the old convention, we'll have more flexibility in this regard as well.

**Comment by bennyguo** [comment]
I agree with Christian. It would be better if we could use something like self.C to update arbitrary values in the dataset. However in the dataset we currently are not able to access the global step, any thoughts?

**Comment by bennyguo** [comment]
And we probably need a more flexible expansion instead of the linear interpolation of the current self.C. For example, in ProlificDreamer, we increase the render resolution at certain steps.

**Comment by voletiv** [comment]
1. The problem indeed was not being able to access global_step in the data, I "fixed" that by making data.collate fn record its own step.
2. radius/fovy start from the eval value, not the higher/lower ones. This is how it was in stable dreamfusion as well
So it's effectively like the (start_iter, start_val, end_val, end_iter) scheme. For example, the fovy range's start and end values change like so:
(0, cfg.eval_fovy, cfg.fovy_range[0], cfg.progressive_until)
(0, cfg.eval_fovy, cfg.fovy_range[1], cfg.progressive_until)

**Comment by bennyguo** [comment]
Let the dataset maintain its state can be problematic when resuming from checkpoints as the state is not saved to the checkpoint. How about setting the global step in update_step?

**Comment by bennyguo** [comment]
You have to make the dataset inherit to Updateable for this to work. See the uncond dataset for an example. One thing to notice is that you also need to disable multi-processing data loading by setting num_workers to 0, or you won't be able to modify self attributes of the dataset (the reason for this is not clear).

**Comment by claforte** [comment]
@voletiv I'd like to try this, this week-end... please include in your description on top an example command line for the 2nd phase of Anya, and a typical result you get (if you have that handy). Thanks!

**Comment by voletiv** [comment]
My typical command for it is:
`python launch.py --config configs/experimental/imagecondition_zero123nerf.yaml --train --gpu 0 system.prompt_processor.prompt="A DSLR 3D photo of a cute anime schoolgirl stands proudly with her arms in the air, pink hair ( unreal engine 5 trending on Artstation Ghibli 4k )" system.weights=outputs/zero123/128_anya_front_rgba.png@20230623-145711/ckpts/last.ckpt system.loggers.wandb.enable=true system.loggers.wandb.project="voletiv-anya-new" data.image_path=./load/images/anya_front_rgba.png system.loggers.wandb.name="anya"`

When I want to switch from SD to DF, I comment/uncomment the relevant lines in the yaml file.

File: threestudio/data/uncond.py [hunk 1]
```diff
@@ -82,42 +92,74 @@ def __init__(self, cfg: Any) -> None:
...
+        self.camera_distance_range = [
+            (1 - r) * self.cfg.eval_camera_distance
+            + r * self.cfg.camera_distance_range[0],
+            (1 - r) * self.cfg.eval_camera_distance
+            + r * self.cfg.camera_distance_range[1],
+        ]
+        self.fovy_range = [
+            (1 - r) * self.cfg.eval_fovy_deg + r * self.cfg.fovy_range[0],
+            (1 - r) * self.cfg.eval_fovy_deg + r * self.cfg.fovy_range[1],
+        ]
```
Comment (claforte) [inline threestudio/data/uncond.py:138]: I recommend we don't vary the camera_distance_range and fovy_range progressively. Changing too many params at once makes it very hard to tune the hyperparams.

Comment (voletiv) [reply threestudio/data/uncond.py]: Done!

**Comment by voletiv** [comment]
I've also added changing of batch_size with resolution_milestone, please check!

File: threestudio/data/image.py [hunk 2]
```diff
@@ -178,6 +179,9 @@ def collate(self, batch) -> Dict[str, Any]:
 
         return batch
 
+    def update_step(self, epoch: int, global_step: int, on_load_weights: bool = False):
```
Comment (bennyguo) [inline threestudio/data/image.py:182]: We actually don't need to do this. If SingleImageIterableDataset is Updateable, all its attributes that are Updateable (like `self.random_pose_generator` here) will be updated automatically. Could you please check this?

File: threestudio/data/uncond.py [hunk 3]
```diff
@@ -82,42 +92,74 @@ def __init__(self, cfg: Any) -> None:
         ]
         self.height: int = self.heights[0]
         self.width: int = self.widths[0]
+        self.batch_size: int = self.batch_sizes[0]
         self.directions_unit_focal = self.directions_unit_focals[0]
+        self.elevation_range = self.cfg.elevation_range
+        self.azimuth_range = self.cfg.azimuth_range
+        self.camera_distance_range = self.cfg.camera_distance_range
+        self.fovy_range = self.cfg.fovy_range
+        self.step = -1
```
Comment (bennyguo) [inline threestudio/data/uncond.py:101]: `self.step` is no longer needed right?

**Review by bennyguo — CHANGES_REQUESTED** [review]

File: DOCUMENTATION.md [hunk 4]
```diff
@@ -28,10 +28,10 @@
...
 | height                 | Union[int, List[int]] | Height of the rendered image in training, should be an integer or a list of integers. If a list of integers, the training height will change according to `resolution_milestones`. Default: 64                                                                                      |
 | width                  | Union[int, List[int]] | Width of the rendered image in training, should be an integer or a list of integers. If a list of integers, the training width will change according to `resolution_milestones`. Default: 64                                                                                        |
+| batch_size             | int                   | Number of images per batch in training. If a list of integers, the batch_size will change according to `resolution_milestones`. Default: 1                                                                                                                                                                                                                                  |
```
Comment (bennyguo) [inline DOCUMENTATION.md:31]: change `int` to `Any`?

**Review by bennyguo — CHANGES_REQUESTED** [review]

File: threestudio/data/image.py (hunk 2 above)
Comment (voletiv) [reply threestudio/data/image.py:182]: I checked it, it didn't work. I had to insert an update_step that called self.random_pose_generator's update_step.

File: threestudio/data/uncond.py (hunk 3 above)
Comment (voletiv) [reply threestudio/data/uncond.py]: Great catch! Removed.

File: DOCUMENTATION.md (hunk 4 above)
Comment (voletiv) [reply DOCUMENTATION.md]: Changed to Union[int, List[int]]

**Comment by bennyguo** [comment]
I'll merge this after the comments have been addressed, but we probably need a better way to control these parameters so that we can support more advanced expansion scheme. Will come up with a proposal later.

**Review by bennyguo — APPROVED** [review]

**Comment by voletiv** [comment]
Agreed!
I didn't want to introduce 4 more parameters (elevation_range_min, elevation_range_max, azimuth_range_min, azimuth_range_max) that could be specified with [start_iter, start_val, end_val, end_iter]. I opted to introduce only one more `progressive_until`. Perhaps there's a better way to deal with this.

**Comment by bennyguo** [comment]
> Agreed! I didn't want to introduce 4 more parameters (elevation_range_min, elevation_range_max, azimuth_range_min, azimuth_range_max) that could be specified with [start_iter, start_val, end_val, end_iter]. I opted to introduce only one more `progressive_until`. Perhaps there's a better way to deal with this.

Yes this is what I thought. Using 4 more parameters sounds like a dumb idea ...

**Pull request merged**